docs/controllers/generics.md

```diff
@@ -0,0 +1,113 @@
+
+This is a useful technique for controllers that need to do _the same thing_ to a bunch of resources, like for instance adding consistent labels / annotations.
+
+### PartialObjectMeta
+
+This is the easiest and cheapest way to create a generic reconciler because it still retains type information, and you do not have to devolve into a fully dynamic api.
+
+You can create generic reconcilers:
+
+```rust
+pub async fn reconcile<K>(obj: Arc<PartialObjectMeta<K>>, ctx: Arc<Context>)
+-> Result<Action>
+where
+    K: Resource<Scope = NamespaceResourceScope, DynamicType = ()>
+        + Clone
+        + DeserializeOwned
+        + Debug,
+{
+    let kind = K::kind(&()).to_string();
+    let ns = obj.namespace().unwrap();
+    let object_name = obj.name_any();
+    let api: Api<PartialObjectMeta<K>> = Api::namespaced(ctx.client, &ns);
+
+    // example work; apply some labels to the object
+    let patch: Patch<serde_json::Value> = get_standard_labels_for(&obj)?;
+    let serverside = PatchParams::apply("labeler");
+    api.patch(&object_name, &serverside, &patch).await?;
+
+    Ok(Action::requeue(Duration::from_secs(5 * 60)))
+}
+```
+
+The generic constraints on the associated type of the [Resource] here means this is a __namespaced__ resource (hence the unwrap). You could remove this bound (or change it for a cluster scoped only bound), but then you could not unwrap.
+
+The `DynamicType = ()` constraint is to indicate that this is one of the normal statically generated api types that we have api information for at the type level (i.e. they come from `k8s-openapi`).
+
+For information about the resource we rely on the generic [Resource] and [ResourceExt] traits which is implemented by [PartialObjectMeta].
+
+!!! note "Diverging Logic"
+
+    You will only get access to metadata of the object doing this. This can be mitigated by doing a `match` on `kind` and creating a more specific `Api<K>` inside a match arm.
+
+
+These reconcilers can be hooked up to a `Controller<K>` with another possibly generic fn.
+
+```rust
+async fn run_controller<K>(client: Client) -> Result<(), Infallible>
+where
+    K: Resource<Scope = NamespaceResourceScope, DynamicType = ()>
+        + Clone
+        + DeserializeOwned
+        + Debug
+        + Sync
+        + Send
+        + 'static,
+{
+    let kind = K::kind(&()).to_string();
+    tracing::info!("Starting controller for {kind}");
+
+    let api = Api::<K>::all(client.clone());
+    let (reader, writer) = reflector::store();
+
+    // controller main stream from metadata_watcher
+    let stream = metadata_watcher(api, watcher::Config::default())
+        .default_backoff()
+        .modify(|x| {
+            x.managed_fields_mut().clear(); // ResourceExt pruning
+        })
+        .reflect(writer)
+        .applied_objects()
+        .predicate_filter(predicates::generation);
```

Comment:
I don't think the filter for generation makes sense here, the controller can only operate on metadata but the filter means it will only act when the spec changes

Reply:
This should actually work, [`generation`](https://docs.rs/kube-runtime/0.86.0/src/kube_runtime/utils/predicate.rs.html#176) reads a metadata property.

Reply:
It's effectively a way to tell the controller to only get notified about spec changes, but you don't get the actual information (because you are only using a `metadata_watcher`)

Reply:
Yeah, but spec changes are irrelevant for this controller, since it can't act on the changes anyways, the reconcile function only has access to `PartialObjectMeta`. Assuming our controller just makes sure there is a certain annotation on the resource, the controller would be triggerd by any spec change, but it would not be triggered if someone removed this annotation.

Reply:
oh, right, that is a good point.
https://github.com/kube-rs/website/pull/47/commits/67780882a488b560db32bc4edbd47dcfe3c3c3a6

Reply:
Thinking about it, maybe using a filter like `predicates::annotations` makes sense?

Reply:
That is a great idea. I'll want to just give that go first in the work controller at some point during the week to verify. Will update the example if it works out well - but in theory that should work.